### ISSUE TITLE
Make hierarchy limits more dynamic

### ISSUE THREAD

**Issue by nilsnolde** [valhalla/valhalla#3868]
ref https://github.com/valhalla/valhalla/issues/3731#issuecomment-1339595535

We could think about making hierarchy limits depend on the density of the input locations. Could also be optional via config and by default use the current values. 

What do you think? Or maybe there are other ways to come up with a better (lower) estimate of how much local/arterial expansion we'll likely need?

**Comment by nilsnolde** [comment]
> Bidirectional A* should already be somewhat dependent on density - it will expand a longer distance (within a hierarchy) in sparse road networks than in dense ones.

Hm, are we talking about the same thing @dnesbitt61? I mean it will expand a longer distance because there's less edges to traverse in a sparse network right? I mean specifically the hierarchy limit's `up_count` (or whatever's called) and/or the distance within locations where we allow transitions. I can't see code in either bidirectional or uni-directional A* that's modifying the hierarchy limits based on density (though there's comments in the latter as you pointed out). I'm not sure if the current limits are like a maximum hierarchy count we can tolerate for having "higher quality" routes or average-ish or rather towards performance and less quality routes with a low limit. Also little feeling if it would provide a better trade-off of performance and route quality in bidirectional. Didn't really think about the impact on unidirectional.

**Comment by nilsnolde** [comment]
> but in a sparse network region it will travel a longer distance before the max transitions are reached

Of course, right! Thanks for explaining, I get what you mean. You're right, it's only a concern for unidirectional.. I'll think about that one then.

**Comment by kevinkreiser** [comment]
> this seems intuitive to me - the count of transitions is the same for dense vs. sparse network regions but in a sparse network region it will travel a longer distance before the max transitions are reached

while i agree that its an intuitive strategy, i just wanted to mention that its not gauranteed that networks are balanced in terms of the number of transitions between heirarchy levels. not only because of tagging conventions (data problems) but also just because networks are somewhat organic creations. i feel like there are locales across the planet where the networks may be similarly dense but vary wildly in terms of number transitions nodes at different levels.

now it may be the case that the counts we set are on the extreme side of things in which case they work for all the variation we find across the planet, if that is the case then we are good. and probably it is the case or we'd hear more about failing or strange routes? anyway it would be interesting to measure the transition density per some small area per level across the globe to see what the extrema are just to see how widely it varies.
